packages/config/src/index.ts

```diff
@@ -0,0 +1,183 @@
+import path from 'path';
+import lodash from 'lodash';
+import findUp from 'find-up';
+import Configstore from 'configstore';
+import TruffleError from '@truffle/error';
+import originalRequire from 'original-require';
+import { getInitialConfig, configProps } from './configDefaults';
+
+const DEFAULT_CONFIG_FILENAME = 'truffle-config.js';
+const BACKUP_CONFIG_FILENAME = 'truffle.js'; // old config filename
+
+class TruffleConfig {
+  [key: string]: any;
+
+  private _deepCopy: string[];
+  private _values: any;
+
+  constructor(
+    truffleDirectory?: string,
+    workingDirectory?: string,
+    network?: any,
+  ) {
+    this._deepCopy = ['compilers'];
+    this._values = getInitialConfig({ truffleDirectory, workingDirectory, network });
+
+    const props = configProps({ configObject: this });
+
+    Object.entries(props)
+      .forEach(([propName, descriptor]) => this.addProp(propName, descriptor));
+  }
+
+  public addProp(propertyName: string, descriptor: any): void {
+    // possible property descriptors
+    //
+    // supports `default` and `transform` in addition to `get` and `set`
+    //
+    // default: specify function to retrieve default value (used by get)
+    // transform: specify function to transform value when (used by set)
+    const self = this;
+
+    Object.defineProperty(this, propertyName, {
+      get:
+        descriptor.get ||
+        function () {
+          // value is specified
+          if (propertyName in self._values) {
+            return self._values[propertyName];
+          }
+
+          // default getter is specified
+          if (descriptor.default) {
+            return descriptor.default();
+          }
+
+          // descriptor is a function
+          return descriptor();
+        },
+      set:
+        descriptor.set ||
+        function (value) {
+          self._values[propertyName] = descriptor.transform
+            ? descriptor.transform(value)
+            : value;
+        },
+      configurable: true,
+      enumerable: true
+    });
+  }
+
+  public normalize(obj: any): any {
```

Comment:
Do I understand correctly that this is just cloning? Can we use Object.assign or the spread operator instead?

Reply:
Looks like it's specifically cloning for the purpose of getting rid of the prototype.

I think `Object.assign` is fine, or the spread operator if you make sure it has the polyfill for Node v8 support.

Reply:
But! Keep `config.normalize` around, even if you don't end up using it, because otherwise it'll  be a breaking change.

Reply:
> But! Keep `Config.normalize` around, even if you don't end up using it, because otherwise it'll be a breaking change.

Yes, I thought over this variant